service.py

```diff
@@ -11,53 +12,190 @@ class DataserviceException(Exception):
+                            yield result
+
+def read_dbgap_xml(accession):
+    """
+    Reads db_gap xml file and fetches consent code and external sample id
+    for a given study
+    :returns: A list of tuples (consent_code, sample_id, consent_name)
+        for each sample in the study.
+    """
+    url = (f'https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/' +
+           f'GetSampleStatus.cgi?study_id={accession}&rettype=xml')
+    data = requests.get(url)
+    if data.status_code != 200:
+        raise DbGapException(f'Request for study {accession} returned non-200 '
+                             f'status code: {data.status_code}')
+
+    data = xmltodict.parse(data.content)
+    study_status = list(dict_or_list('@registration_status', data))
+
+    if study_status[0] in ['released']:
+        dbgap_codes = zip(dict_or_list('@consent_code', data),
+                          dict_or_list('@submitted_sample_id', data),
+                          dict_or_list('@consent_short_name', data))
+        return dbgap_codes
+    else:
+        raise DbGapException(f'study {accession} is not released by dbgap. '
+                             f'registration_status: {study_status[0]}')
 
-def handler(event, context):
+def map_one_study(study, dataservice_api):
+    """
+    Attempt to load a dbGaP xml for a study and call a function for each
+    sample to update
+    :param study: The dbGaP study_id
+    :param consentcode: The name of the function that will be called for each
+        sample to update it inside the dataservice
+    :param dataservice_api: The url of the dataservice api
+    """
+    # Get dbgap released version from dataservice
+    url = f'{dataservice_api}/studies?external_id={study}'
+    resp = requests.get(url)
+
+    # Problem with the request
+    if resp.status_code != 200:
+        raise DataserviceException(f'Problem requesting dataservice: '
+                                   f'{url}, {resp.content}')
+    # There was more than one study returned for this accession code
+    if len(resp.json()['results']) > 1:
+        raise DataserviceException(f'More than one study found for {study}')
+    # There was no study found in the dataservice with the accession code
+    if len(resp.json()['results']) == 0:
+        raise DataserviceException(f'Could not find a study for {study}')
+
+    version = resp.json()['results'][0]['version']
+    # The study has no version registered in the dataservice
+    if not version:
+        raise DataserviceException(f'{study} has no version in dataservice')
+
+    # Need to now invoke new functions in batches to process each sample
+    dbgap_codes = read_dbgap_xml(study+'.'+version)
+    events = []
+    for row in dbgap_codes:
+        events.append(event_generator(study, row))
+    payload = {'Records': events, 'Biospecimens': [], 'GenomicFiles': []}
+    if len(payload['Records']) > 0:
+        handler(payload, DATASERVICE=dataservice_api)
+
+def event_generator(study, row):
```

Comment:
The generators aren't necessary anymore as they are just formatting input for a function call.